src/protocol/payload/tx.rs

```diff
@@ -0,0 +1,868 @@
+                tx.encode(buffer)?;
+            }
+            Tx::V3(tx) => {
+                // The overwintered flag IS set.
+                buffer.write_all(&(3u32 | 1 << 31).to_le_bytes())?;
+                tx.encode(buffer)?;
+            }
+            Tx::V4(tx) => {
+                // The overwintered flag IS set.
+                buffer.write_all(&(4u32 | 1 << 31).to_le_bytes())?;
+                tx.encode(buffer)?;
+            }
+            _ => unimplemented!(),
+        }
+
+        Ok(())
+    }
+
+    pub fn decode(bytes: &mut Cursor<&[u8]>) -> io::Result<Self> {
+        let (version, overwinter) = {
+            const LOW_31_BITS: u32 = (1 << 31) - 1;
+            let header = u32::from_le_bytes(read_n_bytes(bytes)?);
+
+            // Extract transaction version and check if overwinter flag is set.
+            (header & LOW_31_BITS, header >> 31 != 0)
+        };
+
+        let tx = match (version, overwinter) {
+            (1, false) => Self::V1(TxV1::decode(bytes)?),
+            (2, false) => Self::V2(TxV2::decode(bytes)?),
+            (3, true) => Self::V3(TxV3::decode(bytes)?),
+            (4, true) => Self::V4(TxV4::decode(bytes)?),
+            _ => unimplemented!(),
+        };
+
+        Ok(tx)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct TxV1 {
+    tx_in_count: VarInt,
+    tx_in: Vec<TxIn>,
+
+    tx_out_count: VarInt,
+    tx_out: Vec<TxOut>,
+
+    // TODO: newtype?
+    lock_time: u32,
+}
+
+impl TxV1 {
+    fn encode(&self, buffer: &mut Vec<u8>) -> io::Result<()> {
+        self.tx_in_count.encode(buffer)?;
+        for input in &self.tx_in {
+            input.encode(buffer)?;
+        }
+
+        self.tx_out_count.encode(buffer)?;
+        for output in &self.tx_out {
+            output.encode(buffer)?;
+        }
+
+        buffer.write_all(&self.lock_time.to_le_bytes())?;
+
+        Ok(())
+    }
+
+    fn decode(bytes: &mut Cursor<&[u8]>) -> io::Result<Self> {
+        let tx_in_count = VarInt::decode(bytes)?;
+        let mut tx_in = Vec::with_capacity(tx_in_count.0);
+
+        for _ in 0..tx_in_count.0 {
+            let input = TxIn::decode(bytes)?;
+            tx_in.push(input);
+        }
+
+        let tx_out_count = VarInt::decode(bytes)?;
+        let mut tx_out = Vec::with_capacity(tx_out_count.0);
+
+        for _ in 0..tx_out_count.0 {
+            let output = TxOut::decode(bytes)?;
+            tx_out.push(output);
+        }
+
+        let lock_time = u32::from_le_bytes(read_n_bytes(bytes)?);
+
+        Ok(Self {
+            tx_in_count,
+            tx_in,
+            tx_out_count,
+            tx_out,
+            lock_time,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct TxV2 {
+    tx_in_count: VarInt,
+    tx_in: Vec<TxIn>,
+
+    tx_out_count: VarInt,
+    tx_out: Vec<TxOut>,
+
+    lock_time: u32,
+
+    // BCTV14
+    join_split_count: VarInt,
+    join_split: Vec<JoinSplit>,
+
+    // Only present if the join_split count > 0.
```

Comment:
If there is such explicit coupling between these fields, should they not become its own type to encapsulate that? Probably worth the effort since the optional decoding logic is repeating in every version on-wards.

Reply:
Great point! 👍 
There are actually quite a few fields that might require explicit types. In the fields I've introduced in this. Given how long this PR is, I was hesitant to include it. I'll jot this down for a follow-up.